graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+package graphql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+	"strconv"
+
+	"github.com/samsarahq/thunder/reactive"
+)
+
+// ExecutionUnit is a set of execution work that will be done when running
+// a graphql query.  For every source there is an equivalent destination
+// ObjectWriter that is used to record the result of running a section of the
+// graphql query.
+type ExecutionUnit struct {
+	ctx          context.Context
+	field        *Field
+	selection    *Selection
+	sources      []interface{}
+	destinations []*objectWriter
+}
+
+// UnitResolver is a function that executes a function and returns a set of
+// new execution units that need to be run.
+type UnitResolver func(*ExecutionUnit) []*ExecutionUnit
+
+// ExecutionScheduler is an interface that can be provided to the BatchExecutor
+// to control how we traverse the Execution graph.  Examples would include using
+// a bounded goroutine pool, or using unbounded goroutine generation for each
+// execution unit.
+type ExecutionScheduler interface {
+	Run(resolver UnitResolver, startingUnits ...*ExecutionUnit)
+}
+
+func NewBatchExecutor(scheduler ExecutionScheduler) *BatchExecutor {
+	return &BatchExecutor{
+		scheduler: scheduler,
+	}
+}
+
+// BatchExecutor is a GraphQL executor.  Given a query it can run through the
+// execution of the request.
+type BatchExecutor struct {
+	scheduler ExecutionScheduler
+}
+
+// Execute executes a query by traversing the GraphQL query graph and resolving
+// or executing fields.  Any work that needs to be done is passed off to the
+// scheduler to handle managing concurrency of the request.
+// It must return a JSON marshallable response (or an error).
+func (e *BatchExecutor) Execute(ctx context.Context, typ Type, source interface{}, query *Query) (interface{}, error) {
+	queryObject, ok := typ.(*Object)
+	if !ok {
+		return nil, fmt.Errorf("expected query or mutation object for execution, got: %s", typ.String())
+	}
+
+	topLevelSelections := Flatten(query.SelectionSet)
+	topLevelRespWriter := newTopLevelObjectWriter(query.Name)
+	initialSelectionExecUnits := make([]*ExecutionUnit, 0, len(topLevelSelections))
+	writers := make(map[string]*objectWriter)
+	for _, selection := range topLevelSelections {
```

Comment:
could this use executeObject somehow?

Reply:
We'd need to pass in a dummy value as the "source".  (Since we filter out nil sources when unwrapping objects).

Honestly the API is a bit odd because I think we always pass in a nil Source in the top-level `Execute` function.  Maybe we can just get rid of it?  Possibly a follow-up?